fastfield_codecs/src/serialize.rs

```diff
@@ -0,0 +1,245 @@
+    open_from_id(bytes, codec_type)
+}
+
+fn open_codec_from_bytes<C: FastFieldCodec, Item: MonotonicallyMappableToU64>(
+    bytes: OwnedBytes,
+) -> io::Result<Arc<dyn Column<Item>>> {
+    let reader = C::open_from_bytes(bytes)?;
+    Ok(Arc::new(monotonic_map_column(reader, Item::from_u64)))
+}
+
+pub fn open_gcd_from_bytes<WrappedCodec: FastFieldCodec>(
+    bytes: OwnedBytes,
+) -> io::Result<impl Column> {
+    let footer_offset = bytes.len() - 24;
+    let (body, mut footer) = bytes.split(footer_offset);
+    let gcd_params = GCDParams::deserialize(&mut footer)?;
+    let gcd_remap = move |val: u64| gcd_params.min_value + gcd_params.gcd * val;
+    let reader: WrappedCodec::Reader = WrappedCodec::open_from_bytes(body)?;
+    Ok(monotonic_map_column(reader, gcd_remap))
+}
+
+fn open_codec_with_gcd<C: FastFieldCodec, Item: MonotonicallyMappableToU64>(
+    bytes: OwnedBytes,
+) -> io::Result<Arc<dyn Column<Item>>> {
+    let reader = open_gcd_from_bytes::<C>(bytes)?;
+    Ok(Arc::new(monotonic_map_column(reader, Item::from_u64)))
+}
+
+/// Returns correct the reader wrapped in the `DynamicFastFieldReader` enum for the data.
+fn open_from_id<T: MonotonicallyMappableToU64>(
+    mut bytes: OwnedBytes,
+    codec_type: FastFieldCodecType,
+) -> io::Result<Arc<dyn Column<T>>> {
+    match codec_type {
+        FastFieldCodecType::Bitpacked => open_codec_from_bytes::<BitpackedCodec, _>(bytes),
+        FastFieldCodecType::Linear => open_codec_from_bytes::<LinearCodec, _>(bytes),
+        FastFieldCodecType::BlockwiseLinear => {
+            open_codec_from_bytes::<BlockwiseLinearCodec, _>(bytes)
+        }
+        FastFieldCodecType::Gcd => {
+            let codec_type = FastFieldCodecType::deserialize(&mut bytes)?;
+            match codec_type {
+                FastFieldCodecType::Bitpacked => open_codec_with_gcd::<BitpackedCodec, _>(bytes),
+                FastFieldCodecType::Linear => open_codec_with_gcd::<LinearCodec, _>(bytes),
+                FastFieldCodecType::BlockwiseLinear => {
+                    open_codec_with_gcd::<BlockwiseLinearCodec, _>(bytes)
+                }
+                FastFieldCodecType::Gcd => Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Gcd codec wrapped into another gcd codec. This combination is not allowed.",
+                )),
+            }
+        }
+    }
+}
+
+pub fn serialize<T: MonotonicallyMappableToU64>(
+    typed_column: impl Column<T>,
+    output: &mut impl io::Write,
+    codecs: &[FastFieldCodecType],
+) -> io::Result<()> {
+    let column = monotonic_map_column(typed_column, T::to_u64);
+    let gcd_params_opt = if codecs.contains(&FastFieldCodecType::Gcd) {
+        gcd_params(&column)
+    } else {
+        None
+    };
+
+    let gcd_params = if let Some(gcd_params) = gcd_params_opt {
+        gcd_params
+    } else {
+        return serialize_without_gcd(column, output, codecs);
```

Comment:
we could alternatively remove gcd from the codecs and call self

Reply:
I agree! I think we need to remove gcd from the codecs and treat it differently. (somewhat like you use to do).

Have a footer common to all codecs with num_vals, min_val, max_val, and gcd. If gcd == 1 then we return a reader without any gcd mapper.